{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# FROM PREVIOUS\n",
    "\n",
    "# Naive exact matching algorithm\n",
    "def naive(p, t):\n",
    "    occurrences = []\n",
    "    for i in range(len(t) - len(p) + 1):  # loop over alignments\n",
    "        match = True\n",
    "        for j in range(len(p)):  # loop over characters\n",
    "            if t[i+j] != p[j]:  # compare characters\n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurrences.append(i)  # all chars matched; record\n",
    "    return occurrences\n",
    "\n",
    "# Reverse Compliment\n",
    "def reverseComplement(s):\n",
    "    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A', 'N': 'N'}\n",
    "    t = ''\n",
    "    for base in s:\n",
    "        t = complement[base] + t\n",
    "    return t\n",
    "\n",
    "# DNA reference genome from FASTA\n",
    "def readGenome(filename):\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            # ignore header line with genome information\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome\n",
    "\n",
    "# Read/quality strings from FASTQ\n",
    "def readFastq(filename):\n",
    "    sequences = []\n",
    "    qualities = []\n",
    "    with open(filename) as fh:\n",
    "        while True:\n",
    "            fh.readline()  # skip name line\n",
    "            seq = fh.readline().rstrip()  # read base sequence\n",
    "            fh.readline()  # skip placeholder line\n",
    "            qual = fh.readline().rstrip() # base quality line\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "            sequences.append(seq)\n",
    "            qualities.append(qual)\n",
    "    return sequences, qualities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# 1: First, implement a version of the naive exact matching algorithm that is strand-aware. \n",
    "# That is, instead of looking only for occurrences of P in T, additionally look for occurrences of the\n",
    "# reverse complement of P in T. If P is ACT, your function should find occurrences of both ACT\n",
    "# and its reverse complement AGT in T.\n",
    "\n",
    "# If P and its reverse complement are identical (e.g. AACGTT), then a given match offset should be reported only once. \n",
    "# So if your new function is called naive_with_rc, then the old naive function and your new naive_with_rc function \n",
    "# should return the same results when P equals its reverse complement.\n",
    "\n",
    "def naive_with_rc(p, t):\n",
    "    occurences = []\n",
    "    p_rev = reverseComplement(p)\n",
    "    # Loop through alignments\n",
    "    for i in range(len(t) - len(p) + 1):\n",
    "        # Set marker\n",
    "        match = True\n",
    "        # Loop through characters of p\n",
    "        for j in range(len(p)):\n",
    "            if not p[j] == t[i+j]:\n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurences.append(i)\n",
    "        if not p == p_rev:\n",
    "            matchRev = True\n",
    "            for j in range(len(p_rev)):\n",
    "                if not p_rev[j] == t[i+j]:\n",
    "                    matchRev = False\n",
    "                    break\n",
    "            if matchRev:\n",
    "                occurences.append(i)\n",
    "    return occurences\n",
    "\n",
    "p = 'CCC'\n",
    "ten_as = 'AAAAAAAAAA'\n",
    "t = ten_as + 'CCC' + ten_as + 'GGG' + ten_as\n",
    "occurrences = naive_with_rc(p, t)\n",
    "print(occurrences)\n",
    "\n",
    "p = 'CGCG'\n",
    "t = ten_as + 'CGCG' + ten_as + 'CGCG' + ten_as\n",
    "occurrences = naive_with_rc(p, t)\n",
    "print(occurrences)\n",
    "\n",
    "def readGenome(filename):\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            # ignore header line with genome information\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome\n",
    "\n",
    "phix_genome = readGenome('phix.fa')\n",
    "occurrences = naive_with_rc('ATTA', phix_genome)\n",
    "print('offset of leftmost occurrence: %d' % min(occurrences))\n",
    "print('# occurrences: %d' % len(occurrences))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# 2: Next, download and parse the lambda virus genome, at: \n",
    "def readGenome(filename):\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome\n",
    "\n",
    "genome = readGenome('lambda_virus.fa')\n",
    "\n",
    "# Quiz 1. How many times does ğ™°ğ™¶ğ™¶ğšƒ or its reverse complement (ğ™°ğ™²ğ™²ğšƒ) occur in the lambda virus genome? \n",
    "# E.g. if ğ™°ğ™¶ğ™¶ğšƒ occurs 10 times and ğ™°ğ™²ğ™²ğšƒ occurs 12 times, you should report 22.\n",
    "occurences = naive_with_rc('AGGT', genome)\n",
    "print len(occurences)\n",
    "\n",
    "# Quiz 2. How many times does ğšƒğšƒğ™°ğ™° or its reverse complement occur in the lambda virus genome?\n",
    "occurences = naive_with_rc('TTAA', genome)\n",
    "print len(occurences)\n",
    "\n",
    "# Quiz 3. What is the offset of the leftmost occurrence of ğ™°ğ™²ğšƒğ™°ğ™°ğ™¶ğšƒ or its reverse complement in the Lambda virus genome? \n",
    "# E.g. if the leftmost occurrence of ğ™°ğ™²ğšƒğ™°ğ™°ğ™¶ğšƒ is at offset 40 (0-based) and the leftmost occurrence of its reverse \n",
    "# complement ğ™°ğ™²ğšƒğšƒğ™°ğ™¶ğšƒ is at offset 29, then report 29.\n",
    "occurences = naive_with_rc('ACTAAGT', genome)\n",
    "print min(occurences)\n",
    "\n",
    "# Quiz 4. What is the offset of the leftmost occurrence of ğ™°ğ™¶ğšƒğ™²ğ™¶ğ™° or its reverse complement in the Lambda virus genome?\n",
    "occurences = naive_with_rc('AGTCGA', genome)\n",
    "print min(occurences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# 3: As we will discuss, sometimes we would like to find approximate matches for P in T. \n",
    "# That is, we want to find occurrences with one or more differences.\n",
    "# For Questions 5 and 6, make a new version of the ğš—ğšŠğš’ğšŸğš function called ğš—ğšŠğš’ğšŸğš_ğŸ¸ğš–ğš– that allows up to 2 \n",
    "# mismatches per occurrence. \n",
    "# Unlike for the previous questions, do not consider the reverse complement here. \n",
    "# We're looking for approximate matches for P itself, not its reverse complement.\n",
    "\n",
    "def naive_2mm(p, t):\n",
    "    occurrences = []\n",
    "    for i in range(len(t) - len(p) + 1):  # loop over alignments\n",
    "        match = True\n",
    "        mismatch = 0\n",
    "        for j in range(len(p)):  # loop over characters\n",
    "            if t[i+j] != p[j]:  # compare characters\n",
    "                mismatch += 1\n",
    "                if mismatch > 2:\n",
    "                    match = False\n",
    "                    break\n",
    "        if match:\n",
    "            occurrences.append(i)  # all chars matched; record\n",
    "    return occurrences\n",
    "\n",
    "genome = readGenome('lambda_virus.fa')\n",
    "\n",
    "# Quiz 5. How many times does ğšƒğšƒğ™²ğ™°ğ™°ğ™¶ğ™²ğ™² occur in the Lambda virus genome when allowing up to 2 mismatches?\n",
    "occurences = naive_2mm('TTCAAGCC', genome)\n",
    "print len(occurences)\n",
    "\n",
    "# Quiz 6. What is the offset of the leftmost occurrence of ğ™°ğ™¶ğ™¶ğ™°ğ™¶ğ™¶ğšƒğšƒ in the Lambda virus genome when allowing up to 2 mismatches?\n",
    "occurences = naive_2mm('AGGAGGTT', genome)\n",
    "print min(occurences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Quiz 7. \n",
    "# Finally, download and parse the provided FASTQ file containing real DNA sequencing reads derived from a human:\n",
    "!wget https://d28rh4a8wq0iu5.cloudfront.net/ads1/data/ERR037900_1.first1000.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "66"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This dataset has something wrong with it; one of the sequencing cycles is poor quality.\n",
    "\n",
    "# Report which sequencing cycle has the problem. Remember that a sequencing cycle corresponds to a particular \n",
    "# offset in all the reads. For example, if the leftmost read position seems to have a problem consistently across \n",
    "# reads, report 0. If the fourth position from the left has the problem, report 3. \n",
    "# Do whatever analysis you think is needed to identify the bad cycle. \n",
    "# It might help to review the \"Analyzing reads by position\" video.\n",
    "\n",
    "def readFastq(filename):\n",
    "    sequences = []\n",
    "    qualities = []\n",
    "    with open(filename) as fh:\n",
    "        while True:\n",
    "            fh.readline()\n",
    "            seq = fh.readline().rstrip()\n",
    "            fh.readline()\n",
    "            qual = fh.readline().rstrip()\n",
    "            if len(seq) == 0:\n",
    "                break # reached the end\n",
    "            sequences.append(seq)\n",
    "            qualities.append(qual)\n",
    "    return sequences, qualities\n",
    "\n",
    "# Convert Phred33 to Q\n",
    "def phred33ToQ(qual):\n",
    "    return ord(qual) - 33\n",
    "\n",
    "def qualityByPos(reads):\n",
    "    sum_quality = [0] * 100 # sum of quality at each position in read\n",
    "    for read in reads:\n",
    "        for i in range(len(read)):\n",
    "            sum_quality[i] += phred33ToQ(read[i])\n",
    "    return sum_quality.index(min(sum_quality))\n",
    "\n",
    "\n",
    "_, quals = readFastq('ERR037900_1.first1000.fastq')\n",
    "qualityByPos(quals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
